File: projects/heart-linker-app/android/app/src/main/java/com/meishi/app/model/ProfileDraft.kt
package com.meishi.app.model

import android.content.ContentValues
import android.database.Cursor
import com.meishi.app.db.ActiveRecord

/**
 * 編集画面で入力中の内容を一時保存しておくためのキャッシュ。
 * 「保存」または「元に戻す」を押すと消える。
 */
class ProfileDraft(
    override var id: Long = 0,
    var name: String = "",
    var email: String = "",
    var phone: String = "",
    var address: String = "",
    var iconPath: String? = null
) : ActiveRecord() {

    override val tableName: String = "profile_draft"

    override fun toContentValues(): ContentValues = ContentValues().apply {
        put("name", name)
        put("email", email)
        put("phone", phone)
        put("address", address)
        put("icon_path", iconPath)
    }

    fun snsAccounts(): List<SnsAccount> = SnsAccount.findAllForDraft(id)

    fun clear() {
        SnsAccount.deleteAllFor(SnsAccount.TABLE_DRAFT, SnsAccount.COLUMN_DRAFT_ID, id)
        delete()
    }

    companion object {
        private fun fromCursor(c: Cursor): ProfileDraft = ProfileDraft(
            id = c.getLong(c.getColumnIndexOrThrow("id")),
            name = c.getString(c.getColumnIndexOrThrow("name")),
            email = c.getString(c.getColumnIndexOrThrow("email")),
            phone = c.getString(c.getColumnIndexOrThrow("phone")),
            address = c.getString(c.getColumnIndexOrThrow("address")),
            iconPath = c.getString(c.getColumnIndexOrThrow("icon_path"))
        )

        /** 既存の下書きがあれば返す。なければnull */
        fun findExisting(): ProfileDraft? {
            val cursor = rawQuery("SELECT * FROM profile_draft ORDER BY id ASC LIMIT 1")
            cursor.use {
                if (it.moveToFirst()) return fromCursor(it)
            }
            return null
        }

        /** profileの内容から新しい下書きを作る(編集開始時に呼ぶ) */
        fun createFrom(profile: Profile): ProfileDraft {
            val draft = ProfileDraft(
                name = profile.name,
                email = profile.email,
                phone = profile.phone,
                address = profile.address,
                iconPath = profile.iconPath
            )
            draft.save()
            profile.snsAccounts().forEach { sns ->
                SnsAccount.newForDraft(draft.id, sns.type, sns.value, sns.sortOrder).save()
            }
            return draft
        }
    }
}

File: projects/heart-linker-app/ios/Meishi/Models/ProfileDraft.swift
import Foundation

/// 編集中のプロフィールを一時保存するキャッシュ。
/// 「保存」または「元に戻す」で clear() して消す。
final class ProfileDraft {
    var id: Int64 = 0
    var name: String
    var email: String
    var phone: String
    var address: String
    var iconPath: String?

    init(id: Int64 = 0, name: String = "", email: String = "",
         phone: String = "", address: String = "", iconPath: String? = nil) {
        self.id = id; self.name = name; self.email = email
        self.phone = phone; self.address = address; self.iconPath = iconPath
    }

    func snsAccounts() -> [SnsAccount] {
        SnsAccount.findAll(table: "profile_draft_sns", parentColumn: "draft_id", parentId: id)
    }

    @discardableResult
    func save() -> Int64 {
        let db = Database.shared
        let params: [SqlValue] = [.text(name), .text(email), .text(phone), .text(address), .opt(iconPath)]
        if id <= 0 {
            db.run("INSERT INTO profile_draft (name, email, phone, address, icon_path) VALUES (?,?,?,?,?)", params)
            id = db.lastInsertRowId()
        } else {
            db.run("UPDATE profile_draft SET name=?, email=?, phone=?, address=?, icon_path=? WHERE id=\(id)", params)
        }
        return id
    }

    func clear() {
        SnsAccount.deleteAll(table: "profile_draft_sns", parentColumn: "draft_id", parentId: id)
        Database.shared.run("DELETE FROM profile_draft WHERE id=?", [.int(id)])
    }

    // MARK: - Class methods

    static func findExisting() -> ProfileDraft? {
        var result: ProfileDraft?
        Database.shared.query("SELECT * FROM profile_draft ORDER BY id ASC LIMIT 1") { row in
            result = ProfileDraft(id: row.int64(0), name: row.text(1), email: row.text(2),
                                  phone: row.text(3), address: row.text(4), iconPath: row.textOrNil(5))
        }
        return result
    }

    static func createFrom(_ profile: Profile) -> ProfileDraft {
        let draft = ProfileDraft(name: profile.name, email: profile.email,
                                 phone: profile.phone, address: profile.address, iconPath: profile.iconPath)
        draft.save()
        profile.snsAccounts().forEach { sns in
            SnsAccount.forDraft(draftId: draft.id, snsType: sns.snsType, value: sns.value,
                                sortOrder: sns.sortOrder, serviceName: sns.serviceName).save()
        }
        return draft
    }
}
